### ISSUE TITLE
Consider making AzureStorageAttachmentConfiguration settings public

### ISSUE THREAD

**Issue by sorenhansen** [SeanFeldman/ServiceBus.AttachmentPlugin#185]
Consider making the settings that can be provided through the constructors of [`AzureStorageAttachmentConfiguration`](https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/blob/develop/src/ServiceBus.AttachmentPlugin/AzureStorageAttachmentConfiguration.cs) `public` instead of `internal`. This way it's easy to reuse a configuration when implementing clean up.

In particular the following settings would be nice to have access to

- [`ConnectionStringProvider`](https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/blob/08aca98974af5752b8ff01530464e758cf7e0c24/src/ServiceBus.AttachmentPlugin/AzureStorageAttachmentConfiguration.cs#L106)
- [`ContainerName`](https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/blob/08aca98974af5752b8ff01530464e758cf7e0c24/src/ServiceBus.AttachmentPlugin/AzureStorageAttachmentConfiguration.cs#L108
)
- [`MessagePropertyToIdentifyAttachmentBlob`](https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/blob/08aca98974af5752b8ff01530464e758cf7e0c24/src/ServiceBus.AttachmentPlugin/AzureStorageAttachmentConfiguration.cs#L114)

For now it's one solution is to create your own class (like below), that holds the settings, and use that as input to both the `AzureStorageAttachmentConfiguration` and the clean up routine. Making the settings public will eliminate this boilerplate code.

```
public class AzureStorageSettings
{
    public string ConnectionString { get; set; }
    public string ContainerName { get; set; }
    public string MessagePropertyToIdentifyAttachmentBlob { get; set; }
}
```

See Pull Request #189

**Comment by SeanFeldman** [comment]
> This way it's easy to reuse a configuration when implementing clean up.

Could you please describe how is it helping to implement the clean-up?

**Comment by sorenhansen** [comment]
For now I use an instance of my own `AzureStorageSettings` class to hold the settings for both the plugin and my clean-up routine, even though `ContainerName` and `MessagePropertyToIdentifyAttachmentBlob` have the default values. Since they are internal, I can't read them when I need them for cleaning-up.

If these settings were public, I could just use an instance of `AttachmentStorageAttachmentConfiguration`, and read the settings in my clean-up routine.

The clean-up logic in it self, isn't made easier. It's the passing of the required settings that is made a bit simpler.

Hope this clarifies it a bit.

**Comment by SeanFeldman** [comment]
Thank you, that does help.

This is how I understand it at this point. To perform the clean-up you require the following:
1. `ContainerName`
1. `MessagePropertyToIdentifyAttachmentBlob`
1. `ConnectionString`

There are two scenarios: using the defaults and overriding those.

Focusing on `ContainerName` and `MessagePropertyToIdentifyAttachmentBlob` let's assume you're using the defaults. Both properties, `ContainerName` and `MessagePropertyToIdentifyAttachmentBlob` have their defaults provided by the plugin which you'd need to use. Those are [well documented](https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/blob/develop/README.md#configure-blob-container-name) and will **not change** to keep wire compatibility. You can use those strings in your code. My understanding is you don't like to use the strings and rather have something strongly types exposed by the `AzureStorageAttachmentConfiguration` class. That can be done, don't see an issue with it.

The second scenario would be you're providing your own values for  `ContainerName` and `MessagePropertyToIdentifyAttachmentBlob`. In that case, I'd argue your code already owns the values and should hold on to those for the cleanup. It's not the plugin's responsibility to provide your application/system with the configurations. The other way around. 

The same logic would apply to the third parameter, `ConnectionString`. It's under your application's control.

**Comment by SeanFeldman** [comment]
@sorenhansen any thoughts on the last comment?

**Comment by sorenhansen** [comment]
@SeanFeldman Sorry for the late reply, I've been busy with my Easter vacation :-)

Your understanding is right, and you're right about the two scenarios. It's the first one I've run into. I would like the default configuration values to be publicly available.

I can agree with you on the second scanario, where I'm using custom values. Then my code definitely owns the values.

How about having the defaults as public constants then? That way the values are accessible, if needed, but it's up to my application to use them as I see fit?

I've updated the pull request with my suggestions.

**Comment by SeanFeldman** [comment]
@sorenhansen, no worries. Sorry for disturbing you during your holiday.

> How about having the defaults as public constants then? That way the values are accessible, if needed, but it's up to my application to use them as I see fit?

This is exactly the direction I was leading you to 🙂
Happy to hear we've found an understanding. I'll have a look at the changes and if everything is good, release this weekend.

**Comment by sorenhansen** [comment]
@SeanFeldman Thanks for the feedback. I think the end result is better than my original suggestion 👍

**Comment by SeanFeldman** [comment]
@sorenhansen what's your Twitter handler (for credits)?
Also, if you don't mind to share the info about your company or project that uses this plugin for. I'm always happy to add names to the [wall of fame](https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin#whos-trusting-this-plugin-in-production).
If you don't want to do it publicly, you can always ping me directly at feldman.sean @ gmail.com

**Comment by sorenhansen** [comment]
Twitter: @sorenhansen_dk

About company/project, I'll have to get clearance from the top brass, so it will be after Easter.

**Comment by SeanFeldman** [comment]
Closed via 7dfff04543faaa476f73a2d8e49e50b8dee61d5b
